### PR TITLE
Users/jeyou/jenkins download artifacts

### PR DESCRIPTION
**Pull request by jeffyoung**

### CODE REVIEW

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts [hunk 1]
```diff
@@ -21,25 +21,24 @@ class Credential {
...
-                   tl.debug(tl.loc('ServerCallErrorCode', res.statusCode));
-                }
-                if (body) {
-                    tl.debug(body);
+    request.get({url: url, strictSSL: strictSSL}, (err, res, body) => {
+                if (res && body && res.statusCode === 200)  {
+                    defer.resolve(JSON.parse(body));
+                } else {
+                    if (res && res.statusCode) {
+                    tl.debug(tl.loc('ServerCallErrorCode', res.statusCode));
```
Comment (jpricketMSFT) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: needs indentation

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts [hunk 2]
```diff
@@ -20,140 +19,137 @@ class Credential {
...
-    request
-        .get({url: url, strictSSL: strictSSL}, (err, res, body) => {
-            if (res && body && res.statusCode === 200)  {
-                defer.resolve(JSON.parse(body));
-            } else {
-                if (res && res.statusCode) {
-                   tl.debug(tl.loc('ServerCallErrorCode', res.statusCode)); 
+    const defer = Q.defer<any>();
+
+    request.get({url: url, strictSSL: strictSSL}, (err, res, body) => {
```
Comment (bryanmacfarlane) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: Why do we have this function at all?  If you used the typed-rest-client it's one line of code and it's awaitable.  No reason to wrap.  typed-rest-client also supports passing along proxy config from agent etc...

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts [hunk 3]
```diff
@@ -20,140 +19,137 @@ class Credential {
...
-        if (result && result['artifacts']) {
-            let artifacts = result['artifacts'];
-            if (artifacts.length === 0) {
-                defer.reject(new Error(tl.loc('CouldNotFindArtifacts', jobName, jobBuildId)));
+function getArtifactsRelativePaths(serverEndpointUrl: string, jobName: string, jobBuildId: number, cred: Credential, strictSSL: boolean): Q.Promise<string[]> {
+    const defer = Q.defer<string[]>();
+    const artifactQueryUrl: string = `${serverEndpointUrl}/job/${jobName}/${jobBuildId}/api/json?tree=artifacts[*]`;
+
+    getRequest(artifactQueryUrl, cred, strictSSL).then((result) => {
+            if (result && result['artifacts']) {
```
Comment (bryanmacfarlane) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: again

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts [hunk 4]
```diff
@@ -20,140 +19,137 @@ class Credential {
...
-    .then( result => {
-        if (result && result['lastSuccessfulBuild']) {
-            let lastSuccessfulBuildId = result['lastSuccessfulBuild']['id'];
-            if (lastSuccessfulBuildId) {
-                defer.resolve(lastSuccessfulBuildId);
+    const defer = Q.defer<number>();
+    const lastSuccessfulUrl: string = `${serverEndpointUrl}/job/${jobName}/api/json?tree=lastSuccessfulBuild[id,displayname]`;
+
+    getRequest(lastSuccessfulUrl, cred, strictSSL).then((result) => {
+            if (result && result['lastSuccessfulBuild']) {
```
Comment (bryanmacfarlane) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: again

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts (hunk 2 above)
Comment (jeffyoung) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: It's a simple code clean up, not an entire re-factor.

File: Tasks/JenkinsDownloadArtifacts/tslint.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
+{
+  "rules": {
```
Comment (bryanmacfarlane) [inline Tasks/JenkinsDownloadArtifacts/tslint.json]: What causes this tslint.json to be used?  I would prefer that we share a tslint.json in this repo with a way for a specific task to opt out (and perhaps that's the initial setting for all the other tasks).  Then each team does work to fix lint issues and remove the opt out.

What I'm after is consistent engineering in our team

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts (hunk 2 above)
Comment (bryanmacfarlane) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: OK.  We don't have to do it in this PR - just wondering.  Proxy support getting checked in now to task lib.

File: Tasks/JenkinsDownloadArtifacts/tslint.json (hunk 5 above)
Comment (jeffyoung) [inline Tasks/JenkinsDownloadArtifacts/tslint.json]: Changes local to the repo (as in this [pull request](https://github.com/Microsoft/vsts-tasks/pull/4698)) are what causes the tslint.json to be used. Giving each task/team the ability to control their linting rules allows them to manage their code how they want. Lacking a global way to do linting across this repository, that pull request is the suggestion I made. If you want to go another direction, that's fine. Feel free to reject that PR.

**Pull request merged**